src/counter.rs

```diff
@@ -0,0 +1,200 @@
+//! Module holding utilities related to counts
+//! of bytes passed to [AsyncWrite].
+
+use std::task::Poll;
+
+use futures::ready;
+use futures::AsyncWrite;
+use pin_project_lite::pin_project;
+use std::io::{Error, ErrorKind, Result};
+use std::pin::Pin;
+use std::task::Context;
+
+pin_project! {
+    ///An [AsyncWrite] which counts bytes written to the
+    ///wrapped [AsyncWrite].
+    ///```rust
+    /// # use cobalt_async::counter::ByteCounter;
+    /// # use futures::AsyncWriteExt;
+    /// # use futures::io::sink;
+    /// # use tokio_test;
+    /// # tokio_test::block_on(async {
+    ///      let mut counter = ByteCounter::new(sink());
+    ///      counter.write("this is a test".as_bytes()).await?;
+    ///      assert_eq!(14, counter.byte_count());
+    ///      # Ok::<(), anyhow::Error>(())
+    /// # });
+    ///````
+    ///
+    ///## Note
+    ///The count is stored as a [u128] and unchecked addition
+    ///is use to increment the count which means wrapping a
+    ///long running [AsyncWrite] may lead to an overflow.
```

Comment:
```suggestion
    ///long running [AsyncWrite] may lead to an overflow.
```

long running seems like an understatement :-)